pkg/registry/graph.go

```diff
@@ -0,0 +1,326 @@
+	parents := sets.NewString()
+	for parent := range n.workflowParents {
+		parents.Insert(parent.Name())
+	}
+	for parent := range n.chainParents {
+		parents.Insert(parent.Name())
+	}
+	return parents
+}
+
+func (*workflowNode) ParentNames() sets.String { return nil }
+
+// ChildrenNames functions
+
+func (n *nodeWithChildren) ChildrenNames() sets.String {
+	children := sets.NewString()
+	for child := range n.referenceChildren {
+		children.Insert(child.Name())
+	}
+	for child := range n.chainChildren {
+		children.Insert(child.Name())
+	}
+	return children
+}
+
+func (*referenceNode) ChildrenNames() sets.String { return nil }
+
+// AncestorNames functions
+
+func (n *nodeWithParents) AncestorNames() sets.String {
+	ancestors := n.ParentNames()
+	for parent := range n.chainParents {
+		ancestors.Insert(parent.AncestorNames().List()...)
+	}
+	return ancestors
+}
+
+func (*workflowNode) AncestorNames() sets.String { return nil }
+
+// DescendantNames function
+
+func (n *nodeWithChildren) DescendantNames() sets.String {
+	descendants := n.ChildrenNames()
+	for child := range n.chainChildren {
+		descendants.Insert(child.DescendantNames().List()...)
+	}
+	return descendants
+}
+
+func (*referenceNode) DescendantNames() sets.String { return nil }
+
+// Struct helper functions
+// addChild functions
+
+func (n *workflowNode) addChainChild(child *chainNode) {
+	n.chainChildren.insert(child)
+	child.workflowParents.insert(n)
+}
+
+func (n *workflowNode) addReferenceChild(child *referenceNode) {
+	n.referenceChildren.insert(child)
+	child.workflowParents.insert(n)
+}
+
+func (n *chainNode) addChainChild(child *chainNode) {
+	n.chainChildren.insert(child)
+	child.chainParents.insert(n)
+}
+
+func (n *chainNode) addReferenceChild(child *referenceNode) {
+	n.referenceChildren.insert(child)
+	child.chainParents.insert(n)
+}
+
+func newNodeWithName(name string) nodeWithName {
+	return nodeWithName{name: name}
+}
+
+func newNodeWithParents() nodeWithParents {
+	return nodeWithParents{
+		chainParents:    make(chainNodeSet),
+		workflowParents: make(workflowNodeSet),
+	}
+}
+
+func newNodeWithChildren() nodeWithChildren {
+	return nodeWithChildren{
+		chainChildren:     make(chainNodeSet),
+		referenceChildren: make(referenceNodeSet),
+	}
+}
+
+func hasCycles(node *chainNode, ancestors sets.String, traversedPath []string) error {
+	if ancestors == nil {
+		ancestors = make(sets.String)
+	}
+	if ancestors.Has(node.name) {
+		return fmt.Errorf("Cycle detected: %s is an ancestor of itself; traversedPath: %v", node.name, append(traversedPath, node.name))
+	}
+	ancestors.Insert(node.name)
+	for child := range node.chainChildren {
+		if child.Type() != Chain {
+			continue
+		}
+		// get new copy of ancestors and traversedPath so the root node's set isn't changed
+		ancestorsCopy := sets.NewString()
+		for _, ancestor := range ancestors.List() {
+			ancestorsCopy.Insert(ancestor)
+		}
+		traversedPathCopy := append(traversedPath[:0:0], traversedPath...)
```

Comment:
:wizardry:

Reply:
Would it be better to just make a `for` loop to clone the slice? It's slower but more readable. I don't think we'll get datasets so large that performance here will be much of an issue.

Reply:
I think it's fine to leave as is

Reply:
If we cared about performance we'd also switch all the `List()` to `UnsortedList()` and sort post-facto in tests or on display etc where we care